ashura/engine/animation.h

```diff
@@ -2,25 +2,620 @@
 #pragma once
 #include "ashura/std/math.h"
 #include "ashura/std/types.h"
+#include <concepts>
+#include <deque>
+#include <functional>
+#include <map>
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
 
 namespace ash
 {
 
 /// [ ] https://create.roblox.com/docs/ui/animation#style
 /// [ ] Procedural Animation https://www.youtube.com/watch?v=qlfh_rv6khY
 
-// [ ] Keyframes, for time interval x, move from a to b
+// [x] Keyframes, for time interval x, move from a to b
 // [ ] keyframe blending
-// [ ] play
-// [ ] reverse
-// [ ] cancel
+// [x] play
+// [x] reverse
+// [x] cancel
 // [ ] frame-rate customization
-// [ ] move back from point
-// [ ] loop n or forever
+// [x] move back from point
+// [x] loop n or forever
 // [ ] stop after timepoint
 // [ ] alternate
-// [ ] pause
-// [ ] resume in direction
-// [ ] reset animation state to beginning
+// [x] pause
+// [x] resume in direction
+// [x] reset animation state to beginning
+
+// TODO: Support ash Span type instead of using std::string
+
+// Animatable trait applicable to complex types like Color, Vec2, Vec3 etc
+template <typename T>
+concept Animatable = requires(T a, T b, f32 t) {
+  { lerp(a, b, t) } -> std::convertible_to<T>;
+};
+
+namespace animation
+{
+// Lerp implementations for different types
+// Basic arithmetic types
+template <typename T>
+T lerp(const T &start, const T &end, f32 t)
+  requires std::is_arithmetic_v<T>
+{
+  return ash::lerp<T>(start, end, t);
+}
+}        // namespace animation
+
+template <Animatable T>
+struct Animation;
+template <typename T>
+struct Keyframe;
+template <typename T>
+struct Timeline;
+
+struct AnimationTime
+{
+  using Clock     = std::chrono::steady_clock;
+  using TimePoint = Clock::time_point;
+  using Duration  = std::chrono::duration<float>;
+
+  static f32 get_delta(const TimePoint &current, const TimePoint &previous)
+  {
+    return std::chrono::duration_cast<Duration>(current - previous).count();
+  }
+};
+
+enum class CurveType
+{
+  Linear,
```

Comment:
we would eventually need serialization, so please set the enums to specific values, and set the underlying type to a concrete type, i.e. u32